File: python-sdk/frootai/ab_testing.py
"""FrootAI Prompt A/B Testing Framework.

Run prompt experiments across variants, measure quality, pick winners.
Requires a model_fn callback for actual LLM inference.

Usage:
    from frootai.ab_testing import PromptExperiment, PromptVariant

    def my_model(system_prompt: str, query: str) -> str:
        # Call Azure OpenAI, local model, etc.
        return openai_client.chat(system_prompt=system_prompt, query=query)

    def my_scorer(query: str, response: str) -> dict[str, float]:
        return {"groundedness": 4.5, "relevance": 4.0}

    experiment = PromptExperiment(
        name="rag-system-prompt-v2",
        variants=[
            PromptVariant("control", "You are a helpful assistant."),
            PromptVariant("concise", "You are a concise assistant. Answer in 2 sentences max."),
            PromptVariant("expert", "You are an Azure AI expert. Cite sources."),
        ],
        metrics=["groundedness", "relevance", "latency"],
    )

    results = experiment.run(
        test_queries=["What is RAG?", "Explain embeddings"],
        model_fn=my_model,
        scorer_fn=my_scorer,
    )
    winner = experiment.pick_winner(results)
"""

from dataclasses import dataclass, field
from typing import Optional, Callable
import json
import time


@dataclass
class PromptVariant:
    """A single prompt variant in an A/B test."""
    name: str
    system_prompt: str
    weight: float = 1.0


@dataclass
class ExperimentResult:
    """Result of running one variant against one query."""
    variant: str
    query: str
    response: str
    latency_ms: float
    scores: dict[str, float] = field(default_factory=dict)


@dataclass
class PromptExperiment:
    """A/B testing experiment for prompt variants.

    Attributes:
        name: Experiment identifier
        variants: List of prompt variants to test
        metrics: Quality metrics to measure
    """
    name: str
    variants: list[PromptVariant]
    metrics: list[str] = field(default_factory=lambda: ["groundedness", "relevance", "coherence"])

    def run(
        self,
        test_queries: list[str],
        model_fn: Callable[[str, str], str],
        scorer_fn: Optional[Callable[[str, str], dict[str, float]]] = None,
        rounds: int = 1,
    ) -> list[ExperimentResult]:
        """Run the experiment using provided model and scorer functions.

        Args:
            test_queries: Questions to test each variant against.
            model_fn: Callable(system_prompt, query) -> response string.
            scorer_fn: Optional Callable(query, response) -> {metric: score}.
                        If not provided, only latency is measured.
            rounds: Number of rounds to repeat (for statistical stability).
        """
        results = []
        for _ in range(rounds):
            for query in test_queries:
                for variant in self.variants:
                    start = time.perf_counter()
                    response = model_fn(variant.system_prompt, query)
                    latency = (time.perf_counter() - start) * 1000

                    scores = {}
                    if scorer_fn is not None:
                        scores = scorer_fn(query, response)
                    scores["latency_ms"] = round(latency, 1)

                    result = ExperimentResult(
                        variant=variant.name,
                        query=query,
                        response=response,
                        latency_ms=round(latency, 1),
                        scores=scores,
                    )
                    results.append(result)
        return results

    def pick_winner(self, results: list[ExperimentResult]) -> str:
        """Pick the best variant based on average scores (excluding latency)."""
        variant_scores: dict[str, list[float]] = {}
        for r in results:
            if r.variant not in variant_scores:
                variant_scores[r.variant] = []
            quality_scores = {k: v for k, v in r.scores.items() if k != "latency_ms"}
            if quality_scores:
                avg = sum(quality_scores.values()) / len(quality_scores)
                variant_scores[r.variant].append(avg)

        if not variant_scores or all(len(v) == 0 for v in variant_scores.values()):
            # Fall back to lowest latency if no quality scores
            latencies: dict[str, list[float]] = {}
            for r in results:
                latencies.setdefault(r.variant, []).append(r.latency_ms)
            return min(latencies, key=lambda v: sum(latencies[v]) / len(latencies[v]))

        averages = {v: sum(s) / len(s) for v, s in variant_scores.items() if s}
        return max(averages, key=averages.get)

    def summary(self, results: list[ExperimentResult]) -> str:
        """Generate experiment summary."""
        lines = [f"Experiment: {self.name}", "=" * 50]
        variant_data: dict[str, list] = {}
        for r in results:
            variant_data.setdefault(r.variant, []).append(r)

        for variant, data in variant_data.items():
            avg_scores: dict[str, float] = {}
            all_metrics = set()
            for r in data:
                all_metrics.update(r.scores.keys())
            for m in sorted(all_metrics):
                vals = [r.scores[m] for r in data if m in r.scores]
                if vals:
                    avg_scores[m] = sum(vals) / len(vals)
            lines.append(f"\n  Variant: {variant}")
            lines.append(f"  Samples: {len(data)}")
            for m, s in avg_scores.items():
                lines.append(f"    {m}: {s:.2f}")

        winner = self.pick_winner(results)
        lines.append(f"\n  Winner: {winner}")
        return "\n".join(lines)

    def to_json(self) -> str:
        """Export experiment config as JSON."""
        return json.dumps({
            "name": self.name,
            "variants": [{"name": v.name, "system_prompt": v.system_prompt, "weight": v.weight} for v in self.variants],
            "metrics": self.metrics,
        }, indent=2)

File: npm-sdk/src/ab-testing.js
/**
 * FrootAI Prompt A/B Testing Framework
 *
 * Run prompt experiments across variants, measure quality, pick winners.
 * Requires a modelFn callback for actual LLM inference.
 *
 * Usage:
 *   import { PromptExperiment, PromptVariant } from 'frootai';
 *
 *   const experiment = new PromptExperiment({
 *     name: 'rag-system-prompt-v2',
 *     variants: [
 *       new PromptVariant('control', 'You are a helpful assistant.'),
 *       new PromptVariant('concise', 'Answer in 2 sentences max.'),
 *       new PromptVariant('expert', 'You are an Azure AI expert. Cite sources.'),
 *     ],
 *     metrics: ['groundedness', 'relevance', 'latency'],
 *   });
 *
 *   const results = await experiment.run({
 *     queries: ['What is RAG?', 'Explain embeddings'],
 *     modelFn: async (systemPrompt, query) => callLLM(systemPrompt, query),
 *     scorerFn: async (query, response) => ({ groundedness: 4.5, relevance: 4.0 }),
 *   });
 *
 *   const winner = experiment.pickWinner(results);
 */

/**
 * A single prompt variant in an A/B test.
 */
export class PromptVariant {
  constructor(name, systemPrompt, weight = 1.0) {
    this.name = name;
    this.systemPrompt = systemPrompt;
    this.weight = weight;
  }
}

/**
 * Result of running one variant against one query.
 */
export class ExperimentResult {
  constructor({ variant, query, response, latencyMs, scores = {} }) {
    this.variant = variant;
    this.query = query;
    this.response = response;
    this.latencyMs = latencyMs;
    this.scores = scores;
  }
}

/**
 * A/B testing experiment for prompt variants.
 */
export class PromptExperiment {
  /**
   * @param {object} opts
   * @param {string} opts.name - Experiment identifier
   * @param {PromptVariant[]} opts.variants - Variants to test
   * @param {string[]} [opts.metrics] - Quality metrics to measure
   */
  constructor({ name, variants, metrics = ['groundedness', 'relevance', 'coherence'] }) {
    this.name = name;
    this.variants = variants;
    this.metrics = metrics;
  }

  /**
   * Run the experiment using provided model and scorer functions.
   * @param {object} opts
   * @param {string[]} opts.queries - Test queries
   * @param {function} opts.modelFn - async (systemPrompt, query) => response
   * @param {function} [opts.scorerFn] - async (query, response) => {metric: score}
   * @param {number} [opts.rounds=1] - Rounds to repeat for statistical stability
   * @returns {Promise<ExperimentResult[]>}
   */
  async run({ queries, modelFn, scorerFn = null, rounds = 1 }) {
    const results = [];

    for (let r = 0; r < rounds; r++) {
      for (const query of queries) {
        for (const variant of this.variants) {
          const start = performance.now();
          const response = await modelFn(variant.systemPrompt, query);
          const latencyMs = Math.round((performance.now() - start) * 10) / 10;

          let scores = {};
          if (scorerFn) {
            scores = await scorerFn(query, response);
          }
          scores.latency_ms = latencyMs;

          results.push(new ExperimentResult({
            variant: variant.name,
            query,
            response,
            latencyMs,
            scores,
          }));
        }
      }
    }

    return results;
  }

  /**
   * Pick the best variant based on average scores (excluding latency).
   * Falls back to lowest latency if no quality scores available.
   * @param {ExperimentResult[]} results
   * @returns {string} Winner variant name
   */
  pickWinner(results) {
    const variantScores = {};

    for (const r of results) {
      if (!variantScores[r.variant]) variantScores[r.variant] = [];
      const qualityScores = Object.entries(r.scores)
        .filter(([k]) => k !== 'latency_ms')
        .map(([, v]) => v);
      if (qualityScores.length > 0) {
        const avg = qualityScores.reduce((a, b) => a + b, 0) / qualityScores.length;
        variantScores[r.variant].push(avg);
      }
    }

    const hasScores = Object.values(variantScores).some(v => v.length > 0);
    if (!hasScores) {
      // Fall back to lowest latency
      const latencies = {};
      for (const r of results) {
        if (!latencies[r.variant]) latencies[r.variant] = [];
        latencies[r.variant].push(r.latencyMs);
      }
      let bestVariant = null, bestAvg = Infinity;
      for (const [v, vals] of Object.entries(latencies)) {
        const avg = vals.reduce((a, b) => a + b, 0) / vals.length;
        if (avg < bestAvg) { bestAvg = avg; bestVariant = v; }
      }
      return bestVariant;
    }

    let bestVariant = null, bestAvg = -Infinity;
    for (const [v, scores] of Object.entries(variantScores)) {
      if (scores.length === 0) continue;
      const avg = scores.reduce((a, b) => a + b, 0) / scores.length;
      if (avg > bestAvg) { bestAvg = avg; bestVariant = v; }
    }
    return bestVariant;
  }

  /**
   * Generate a human-readable experiment summary.
   * @param {ExperimentResult[]} results
   * @returns {string}
   */
  summary(results) {
    const lines = [`Experiment: ${this.name}`, '='.repeat(50)];
    const variantData = {};

    for (const r of results) {
      if (!variantData[r.variant]) variantData[r.variant] = [];
      variantData[r.variant].push(r);
    }

    for (const [variant, data] of Object.entries(variantData)) {
      const allMetrics = new Set();
      for (const r of data) Object.keys(r.scores).forEach(k => allMetrics.add(k));

      const avgScores = {};
      for (const m of [...allMetrics].sort()) {
        const vals = data.filter(r => m in r.scores).map(r => r.scores[m]);
        if (vals.length > 0) avgScores[m] = vals.reduce((a, b) => a + b, 0) / vals.length;
      }

      lines.push(`\n  Variant: ${variant}`);
      lines.push(`  Samples: ${data.length}`);
      for (const [m, s] of Object.entries(avgScores)) {
        lines.push(`    ${m}: ${s.toFixed(2)}`);
      }
    }

    const winner = this.pickWinner(results);
    lines.push(`\n  Winner: ${winner}`);
    return lines.join('\n');
  }

  /**
   * Export experiment config as JSON string.
   * @returns {string}
   */
  toJSON() {
    return JSON.stringify({
      name: this.name,
      variants: this.variants.map(v => ({
        name: v.name, systemPrompt: v.systemPrompt, weight: v.weight,
      })),
      metrics: this.metrics,
    }, null, 2);
  }
}
